// belle/lib/index.js
'use strict';

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _componentsButton = require('./components/Button');

var _componentsButton2 = _interopRequireDefault(_componentsButton);

var _componentsCard = require('./components/Card');

var _componentsCard2 = _interopRequireDefault(_componentsCard);

var _componentsChoice = require('./components/Choice');

var _componentsChoice2 = _interopRequireDefault(_componentsChoice);

var _componentsOption = require('./components/Option');

var _componentsOption2 = _interopRequireDefault(_componentsOption);

var _componentsPlaceholder = require('./components/Placeholder');

var _componentsPlaceholder2 = _interopRequireDefault(_componentsPlaceholder);

var _componentsSelect = require('./components/Select');

var _componentsSelect2 = _interopRequireDefault(_componentsSelect);

var _componentsSeparator = require('./components/Separator');

var _componentsSeparator2 = _interopRequireDefault(_componentsSeparator);

var _componentsTextInput = require('./components/TextInput');

var _componentsTextInput2 = _interopRequireDefault(_componentsTextInput);

var _componentsRating = require('./components/Rating');

var _componentsRating2 = _interopRequireDefault(_componentsRating);

var _componentsComboBox = require('./components/ComboBox');

var _componentsComboBox2 = _interopRequireDefault(_componentsComboBox);

var _componentsToggle = require('./components/Toggle');

var _componentsToggle2 = _interopRequireDefault(_componentsToggle);

var _styleButton = require('./style/button');

var _styleButton2 = _interopRequireDefault(_styleButton);

var _styleCard = require('./style/card');

var _styleCard2 = _interopRequireDefault(_styleCard);

var _stylePlaceholder = require('./style/placeholder');

var _stylePlaceholder2 = _interopRequireDefault(_stylePlaceholder);

var _styleOption = require('./style/option');

var _styleOption2 = _interopRequireDefault(_styleOption);

var _styleSelect = require('./style/select');

var _styleSelect2 = _interopRequireDefault(_styleSelect);

var _styleSeparator = require('./style/separator');

var _styleSeparator2 = _interopRequireDefault(_styleSeparator);

var _styleTextInput = require('./style/text-input');

var _styleTextInput2 = _interopRequireDefault(_styleTextInput);

var _styleToggle = require('./style/toggle');

var _styleToggle2 = _interopRequireDefault(_styleToggle);

var _styleRating = require('./style/rating');

var _styleRating2 = _interopRequireDefault(_styleRating);

var _styleComboBox = require('./style/combo-box');

var _styleComboBox2 = _interopRequireDefault(_styleComboBox);

var _configSelect = require('./config/select');

var _configSelect2 = _interopRequireDefault(_configSelect);

var _configButton = require('./config/button');

var _configButton2 = _interopRequireDefault(_configButton);

var _configRating = require('./config/rating');

var _configRating2 = _interopRequireDefault(_configRating);

var _configToggle = require('./config/toggle');

var _configToggle2 = _interopRequireDefault(_configToggle);

module.exports = {
  Button: _componentsButton2['default'],
  Card: _componentsCard2['default'],
  Choice: _componentsChoice2['default'],
  Option: _componentsOption2['default'],
  Placeholder: _componentsPlaceholder2['default'],
  Select: _componentsSelect2['default'],
  Separator: _componentsSeparator2['default'],
  TextInput: _componentsTextInput2['default'],
  Rating: _componentsRating2['default'],
  ComboBox: _componentsComboBox2['default'],
  Toggle: _componentsToggle2['default'],
  style: {
    button: _styleButton2['default'],
    card: _styleCard2['default'],
    option: _styleOption2['default'],
    placeholder: _stylePlaceholder2['default'],
    rating: _styleRating2['default'],
    select: _styleSelect2['default'],
    separator: _styleSeparator2['default'],
    textInput: _styleTextInput2['default'],
    toggle: _styleToggle2['default'],
    comboBox: _styleComboBox2['default']
  },
  config: {
    select: _configSelect2['default'],
    button: _configButton2['default'],
    rating: _configRating2['default'],
    toggle: _configToggle2['default']
  }
};
