// belle/lib/utils/helpers.js
// The following helper functions are based on Underscore.js

/**
 * Returns true if the object contain the given key.
 *
 * @param {object} obj - object to be inspected
 * @param {string} key - name of the property
 */
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.has = has;
exports.omit = omit;
exports.filter = filter;
exports.map = map;
exports.find = find;
exports.isEmpty = isEmpty;
exports.findIndex = findIndex;
exports.first = first;
exports.last = last;
exports.size = size;
exports.some = some;
exports.union = union;
exports.uniqueId = uniqueId;
exports.extend = extend;
exports.flatten = flatten;

function has(obj, key) {
  return obj !== undefined && obj !== null && Object.prototype.hasOwnProperty.call(obj, key);
}

/**
 * Return a copy of the object, filtered to omit the blacklisted keys (or array of keys).
 *
 * @param {object} obj - object the returned object is based on
 * @param {string|string[]} fields - the key or list of keys of the property to omit
 */

function omit(obj, fields) {
  if (obj) {
    var result = {};
    for (var key in obj) {
      if (obj.hasOwnProperty(key) && (!fields || fields.indexOf(key) < 0)) {
        result[key] = obj[key];
      }
    }
    return result;
  }
}

/**
 * Looks through each value in the list, returning an array of all the values
 * that pass a truth test (predicate).
 *
 * @param {array} iterable - the iterable object to be filtered
 * @param {function} predicate - function returning true when provided with an entry as argument
 * @param {object} [context] - context for the predicate function call
 */

function filter(iterable, predicate, context) {
  if (iterable) {
    var _ret = (function () {
      var result = [];
      iterable.forEach(function (obj) {
        if (predicate && predicate.call(context, obj)) {
          result.push(obj);
        }
      });
      return {
        v: result
      };
    })();

    if (typeof _ret === 'object') return _ret.v;
  }
}

/**
 * Returns true if the provided object is an array.
 *
 * @param {object} obj - object to be inspected
 */
function isArrayLike(obj) {
  var length = obj.length;
  return typeof length === 'number' && length >= 0;
}

/**
 * Returns all the names of the object's properties.
 *
 * @param {object} obj - object to be used
 */
function keys(obj) {
  var objKeys = [];
  for (var key in obj) {
    if (has(obj, key)) objKeys.push(key);
  }return objKeys;
}

/**
 * Returns a new array of values by mapping each value in list through a transformation function (predicate). If object is a not an array, predicate's arguments will be (value, key, list).
 *
 * @param {object|array} obj - object to be based upon
 * @param {function} predicate - function returning the a new version of the entry
 * @param {object} [context] - context for the predicate function call
 */

function map(obj, predicate, context) {
  if (obj) {
    var result = [];
    var objKeys = !isArrayLike(obj) && keys(obj);
    var _length = (objKeys || obj).length;
    for (var index = 0; index < _length; index++) {
      var currentKey = objKeys ? objKeys[index] : index;
      if (predicate) {
        result[index] = predicate.call(context, obj[currentKey], currentKey);
      }
    }
    return result;
  }
}

/**
 * Returns the first value that passes a truth test (predicate), or undefined if
 * no value passes the test. Only works for iterable objects e.g. arrays.
 *
 * @param {array} iterable - the iterable object to be searched
 * @param {function} predicate - function returning true in case of a positive match
 * @param {object} [context] - context for the predicate function call
 */

function find(iterable, predicate, context) {
  if (iterable) {
    var result = undefined;
    for (var index = 0; index < iterable.length; index++) {
      if (predicate && predicate.call(context, iterable[index])) {
        result = iterable[index];
        break;
      }
    }
    return result;
  }
}

/**
 * Returns true if object contains no values (no enumerable own-properties).
 *
 * @param {array} iterable - an iterable object
 */

function isEmpty(iterable) {
  return !iterable || iterable.length === 0;
}

/**
 * Returns the index of the first value that passes a truth test (predicate), or undefined if
 * no value passes the test. Only works for iterable objects e.g. arrays.
 *
 * @param {array} iterable - the iterable object to be searched
 * @param {function} predicate - function returning true in case of a positive match
 * @param {object} [context] - context for the predicate function call
 */

function findIndex(iterable, predicate, context) {
  if (iterable) {
    var result = undefined;
    for (var index = 0; index < iterable.length; index++) {
      if (predicate && predicate.call(context, iterable[index])) {
        result = index;
        break;
      }
    }
    return result;
  }
}

/**
 * Returns the first element of an iterable object.
 *
 * @param {array} iterable - must be an iterable object
 */

function first(iterable) {
  if (iterable && iterable.length > 0) {
    return iterable[0];
  }
}

/**
 * Returns the last element of an iterable object.
 *
 * @param {array} iterable - must be an iterable object
 */

function last(iterable) {
  if (iterable && iterable.length > 0) {
    return iterable[iterable.length - 1];
  }
}

/**
 * Return the number of values in the list.
 *
 * @param {array} iterable - must be an iterable object
 */

function size(iterable) {
  if (iterable) {
    return iterable.length;
  }
  return 0;
}

/**
 * Returns true if any of the values in the list pass the predicate truth test.
 *
 * @param {array} iterable - iterable object to be searched
 * @param {function} predicate - function returning true in case of a positive match
 * @param {object} [context] - context for the predicate function call
 */

function some(iterable, predicate, context) {
  if (iterable) {
    var result = undefined;
    for (var index = 0; index < iterable.length; index++) {
      if (predicate && predicate.call(context, iterable[index])) {
        result = true;
        break;
      }
    }
    return result;
  }
}

/**
 * Returns the union of the passed-in arrays: the list of unique items, in order, that are present in one or more of the arrays.
 *
 * @param {...array} arrs - at least two iterable objects must be provide
 */

function union() {
  for (var _len = arguments.length, arrs = Array(_len), _key = 0; _key < _len; _key++) {
    arrs[_key] = arguments[_key];
  }

  if (arrs) {
    var _ret2 = (function () {
      var result = [];
      arrs.forEach(function (arr) {
        if (arr) {
          arr.forEach(function (obj) {
            if (result.indexOf(obj) < 0) {
              result.push(obj);
            }
          });
        }
      });
      return {
        v: result
      };
    })();

    if (typeof _ret2 === 'object') return _ret2.v;
  }
}

var idCounter = 0;

/**
 * Generate a globally-unique id.
 *
 * @param {string} [prefix] - if prefix is passed, the id will be appended to it.
 */

function uniqueId(prefix) {
  var id = ++idCounter + '';
  return prefix ? prefix + id : id;
}

/**
 * Copy all of the properties in the source objects over to the destination object, and return the destination object. It's in-order, so the last source will override properties of the same name in previous arguments.
 *
 * @param {object} obj1 - object to be extended
 * @param {...object} objs - at least one but optionally more objects an be provided
 */

function extend(obj1) {
  for (var _len2 = arguments.length, objs = Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
    objs[_key2 - 1] = arguments[_key2];
  }

  if (obj1 && objs) {
    objs.forEach(function (obj) {
      if (obj) {
        for (var key in obj) {
          if (obj.hasOwnProperty(key)) {
            obj1[key] = obj[key];
          }
        }
      }
    });
  }
  return obj1;
}

/**
 * Recursive function for flattening an iterable.
 *
 * @param {object} output - base object to be updated
 * @param {object} element - input object to be merged into the output
 */
function flattenInternal(output, element) {
  if (element) {
    element.forEach(function (obj) {
      if (Array.isArray(obj)) {
        flattenInternal(output, obj);
      } else {
        output.push(obj);
      }
    });
  }
}

/**
 * Flattens a nested array (the nesting can be to any depth).
 *
 * @param {...array} arrays - at least one array must be provided
 */

function flatten() {
  for (var _len3 = arguments.length, arrays = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
    arrays[_key3] = arguments[_key3];
  }

  if (arrays) {
    var result = [];
    flattenInternal(result, arrays);
    return result;
  }
}
