// belle/lib/utils/inject-style.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.injectStyles = injectStyles;
exports.removeStyle = removeStyle;
exports.removeAllStyles = removeAllStyles;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _utilsHelpers = require('../utils/helpers');

var _reactLibCSSPropertyOperations = require('react/lib/CSSPropertyOperations');

var _reactLibCSSPropertyOperations2 = _interopRequireDefault(_reactLibCSSPropertyOperations);

var _reactLibExecutionEnvironment = require('react/lib/ExecutionEnvironment');

var _styleAnimations = require('../style/animations');

var _styleAnimations2 = _interopRequireDefault(_styleAnimations);

var styleElement = undefined;
var styleStorage = {};

/**
 * Injects the provided style into the styleStore.
 */
function injectStyleTag() {
  if (!styleElement && _reactLibExecutionEnvironment.canUseDOM) {
    styleElement = document.createElement('style');
    document.body.appendChild(styleElement);
    styleElement.setAttribute('class', 'belle-style');
  }
}

/**
 * Injects the provided style into the styleStore.
 */
function updateStore(styleId, style, pseudoClass, disabled) {
  styleStorage[styleId] = styleStorage[styleId] || {};
  if (disabled) {
    styleStorage[styleId].disabledPseudoClasses = styleStorage[styleId].disabledPseudoClasses || {};
    styleStorage[styleId].disabledPseudoClasses[pseudoClass] = style;
  } else {
    styleStorage[styleId].pseudoClasses = styleStorage[styleId].pseudoClasses || {};
    styleStorage[styleId].pseudoClasses[pseudoClass] = style;
  }
}

/**
 * Constructs all the stored styles & injects them to the DOM.
 */
function createMarkupOnPseudoClass(pseudoClasses, id, disabled) {
  return (0, _utilsHelpers.map)(pseudoClasses, function (style, pseudoClass) {
    var styleString = _reactLibCSSPropertyOperations2['default'].createMarkupForStyles(style);
    var styleWithImportant = styleString.replace(/;/g, ' !important;');

    return disabled ? '.' + id + '[disabled]:' + pseudoClass + ' {' + styleWithImportant + '}' : '.' + id + ':' + pseudoClass + ' {' + styleWithImportant + '}';
  });
}

function updateStyling() {
  var styles = (0, _utilsHelpers.map)(styleStorage, function (storageEntry, id) {
    var pseudoClassesArray = [];

    if (storageEntry.pseudoClasses) {
      pseudoClassesArray.push(createMarkupOnPseudoClass(storageEntry.pseudoClasses, id, false));
    }
    if (storageEntry.disabledPseudoClasses) {
      pseudoClassesArray.push(createMarkupOnPseudoClass(storageEntry.disabledPseudoClasses, id, true));
    }

    return pseudoClassesArray;
  });
  if (styleElement) {
    styleElement.innerHTML = (0, _utilsHelpers.flatten)([_styleAnimations2['default'], styles]).join(' ');
  }
}

/**
 * Injects a style tag and adds multiple passed styles.
 *
 * By using this function someone can make sure the DOM is updated only once.
 *
 * @example
 * ```
 * const styles = [
 *   {
 *     id: 'style-0.0.2',
 *     style: { color: '#F00' },
 *     pseudoClass: 'hover'
 *   }
 * ];
 * injectStyles(styles);
 * ```
 */

function injectStyles(styles) {
  injectStyleTag();
  styles.forEach(function (style) {
    updateStore(style.id, style.style, style.pseudoClass, style.disabled);
  });
  updateStyling();
}

/**
 * Removes all pseudoClass styles based on the provided styleId.
 */

function removeStyle(styleId) {
  delete styleStorage[styleId];
  updateStyling();
}

/**
 * Removes all pseudoClass styles based on all provided styleIds.
 */

function removeAllStyles(styleIds) {
  styleIds.forEach(function (styleId) {
    delete styleStorage[styleId];
  });
  updateStyling();
}

/**
 * Injects a style tag and adds the passed style for the provided pseudoClass.
 */

exports['default'] = function (styleId, style, pseudoClass, disabled) {
  injectStyleTag();
  updateStore(styleId, style, pseudoClass, disabled);
  updateStyling();
};
