// belle/lib/components/Placeholder.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _utilsHelpers = require('../utils/helpers');

var _stylePlaceholder = require('../style/placeholder');

var _stylePlaceholder2 = _interopRequireDefault(_stylePlaceholder);

/**
 * Returns an object with properties that are relevant for the wrapping div.
 */
function sanitizeChildProps(properties) {
  return (0, _utilsHelpers.omit)(properties, ['style', 'disabledStyle', '_isDisabled']);
}

/**
 * Placeholder component.
 *
 * This component should be used together with Belle's Select.
 */

var Placeholder = (function (_Component) {
  function Placeholder(properties) {
    _classCallCheck(this, Placeholder);

    _get(Object.getPrototypeOf(Placeholder.prototype), 'constructor', this).call(this, properties);
    this.state = {
      childProps: sanitizeChildProps(properties)
    };
  }

  _inherits(Placeholder, _Component);

  _createClass(Placeholder, [{
    key: 'componentWillReceiveProps',

    /**
     * Update the childProps based on the updated properties passed to the
     * Placeholder.
     */
    value: function componentWillReceiveProps(properties) {
      this.setState({ childProps: sanitizeChildProps(properties) });
    }
  }, {
    key: 'render',
    value: function render() {
      var computedStyle = (0, _utilsHelpers.extend)({}, _stylePlaceholder2['default'].style, this.props.style);
      if (this.props._isDisabled) {
        computedStyle = (0, _utilsHelpers.extend)(computedStyle, _stylePlaceholder2['default'].disabledStyle, this.props.disabledStyle);
      }

      return _react2['default'].createElement(
        'div',
        _extends({ style: computedStyle }, this.state.childProps),
        this.props.children
      );
    }
  }], [{
    key: 'displayName',
    value: 'Belle Placeholder',
    enumerable: true
  }, {
    key: 'propTypes',
    value: {
      children: _react2['default'].PropTypes.oneOfType([_react2['default'].PropTypes.arrayOf(_react2['default'].PropTypes.node), _react2['default'].PropTypes.node]),
      style: _react2['default'].PropTypes.object,
      disabledStyle: _react2['default'].PropTypes.object,
      _isDisabled: _react2['default'].PropTypes.bool
    },
    enumerable: true
  }, {
    key: 'defaultProps',
    value: {
      _isDisabled: false
    },
    enumerable: true
  }]);

  return Placeholder;
})(_react.Component);

exports['default'] = Placeholder;
module.exports = exports['default'];
